### ISSUE TITLE
학습방법이 궁금합니다.

### ISSUE THREAD

**Issue by kht6163** [KoJLabs/KoTAN#5]
혹시 nllb fine tuning 하는 방법 질문드려도 괜찮나요?

**Comment by merry555** [comment]
먼저 저희 KoTAN에 관심 가저주셔서 감사합니다.

회사 내부에서 개발했기 때문에 nllb fine tuning하는 전체 소스코드는 공개가 불가능 합니다. 하지만 궁금하신 점이 있다면 질문주시면 답변드리겠습니다.

**Comment by kht6163** [comment]
아 혹시 그러면 제가 사용하는 코드에대한 조언은 혹시 괜찮을까요?

**Comment by merry555** [comment]
> 아 혹시 그러면 제가 사용하는 코드에대한 조언은 혹시 괜찮을까요?

네, 사용하신 코드 보여주시면 조언 드리겠습니다.

**Comment by kht6163** [comment]
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM, Seq2SeqTrainingArguments, Seq2SeqTrainer, NllbTokenizerFast
import pandas as pd
from sklearn.model_selection import train_test_split
import torch
from torch.nn.parallel import DistributedDataParallel as DDP
import torch.utils.data
from transformers import DataCollatorForSeq2Seq
import evaluate
import numpy as np
from argparse import ArgumentParser
import multiprocessing

# from peft import LoraConfig, TaskType
# import peft

device = 'cuda' if torch.cuda.is_available() else 'cpu'


 tokenizer = AutoTokenizer.from_pretrained(
     "facebook/nllb-200-distilled-600M", src_lang="kor_Hang", tgt_lang="eng_Latn"
 )

model = AutoModelForSeq2SeqLM.from_pretrained("facebook/nllb-200-distilled-600M")

# peft_config = LoraConfig(
#     target_modules=['q_proj','v_proj'], task_type=TaskType.SEQ_2_SEQ_LM, inference_mode=False, r=8, lora_alpha=32, lora_dropout=0.1
# )

# model = peft.get_peft_model(model, peft_config)
# model.print_trainable_parameters()

from datasets import Dataset
# import json
# with open('./kor-eng_news.json','r')as f:
#     a = json.load(f)

train_data = pd.read_csv('./train-ko-en.csv', names=['kor_Hang', 'eng_Latn'])

dataset = Dataset.from_pandas(train_data)
dataset = dataset.train_test_split(test_size=0.1, shuffle=True)

def preprocess_function(examples):
    model_inputs = tokenizer(examples['kor_Hang'], text_target=examples['eng_Latn'], max_length=512, truncation=True)
    return model_inputs

# NUM_CPU = int(multiprocessing.cpu_count() / 2)

data_collator = DataCollatorForSeq2Seq(tokenizer=tokenizer, model=model)
tokenized_datasets = dataset.map(preprocess_function, batched=True)

import numpy as np
import evaluate
metric = evaluate.load("sacrebleu")

def postprocess_text(preds, labels):
    preds = [pred.strip() for pred in preds]
    labels = [[label.strip()] for label in labels]
    return preds, labels

def compute_metrics(eval_preds):
    preds, labels = eval_preds
    if isinstance(preds, tuple):
        preds = preds[0]
    decoded_preds = tokenizer.batch_decode(preds, skip_special_tokens=True)

    labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
    decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)

    decoded_preds, decoded_labels = postprocess_text(decoded_preds, decoded_labels)

    result = metric.compute(predictions=decoded_preds, references=decoded_labels)
    result = {"bleu": result["score"]}

    prediction_lens = [np.count_nonzero(pred != tokenizer.pad_token_id) for pred in preds]
    result["gen_len"] = np.mean(prediction_lens)
    result = {k: round(v, 4) for k, v in result.items()}
    return result

# training_args = Seq2SeqTrainingArguments(
#     output_dir='test1',
#     evaluation_strategy="steps",
#     save_strategy='steps',
#     logging_steps=1,
#     eval_steps=10000,
#     save_steps=10000,
#     learning_rate=2e-5,
#     per_device_train_batch_size=16,
#     per_device_eval_batch_size=16,
#     weight_decay=0.01,
#     num_train_epochs=10,
#     predict_with_generate=True,
#     load_best_model_at_end=True,
#     save_total_limit=5,
#     metric_for_best_model='eval_bleu',
#     report_to="wandb",
#     run_name="l4x4 gcp",
# )

training_args = Seq2SeqTrainingArguments(
    output_dir="chkpt3",
    learning_rate=2e-5,
    logging_dir='logs1',
    logging_steps=1,
    weight_decay=0.01,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    num_train_epochs=5,
    save_steps=10000,
    save_total_limit=50,
    evaluation_strategy="epoch",
    predict_with_generate=True,
)

trainer = Seq2SeqTrainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["test"],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)

# print(trainer.evaluate())
trainer.train()
trainer.save_model("result3")

이렇게 사용하고 있습니다.

**Comment by kht6163** [comment]
from transformers import AutoTokenizer, AutoModelForSeq2SeqLM, Seq2SeqTrainingArguments, Seq2SeqTrainer, NllbTokenizerFast
import pandas as pd
from sklearn.model_selection import train_test_split
import torch
from torch.nn.parallel import DistributedDataParallel as DDP
import torch.utils.data
from transformers import DataCollatorForSeq2Seq
import evaluate
import numpy as np
from argparse import ArgumentParser
import multiprocessing


tokenizer = AutoTokenizer.from_pretrained(
     "facebook/nllb-200-distilled-600M", src_lang="kor_Hang", tgt_lang="eng_Latn"
)
model = AutoModelForSeq2SeqLM.from_pretrained("facebook/nllb-200-distilled-600M")

from datasets import Dataset

train_data = pd.read_csv('./train-ko-en.csv', names=['kor_Hang', 'eng_Latn'])

dataset = Dataset.from_pandas(train_data)
dataset = dataset.train_test_split(test_size=0.1, shuffle=True)

def preprocess_function(examples):
    model_inputs = tokenizer(examples['kor_Hang'], text_target=examples['eng_Latn'], max_length=512, truncation=True)
    return model_inputs

data_collator = DataCollatorForSeq2Seq(tokenizer=tokenizer, model=model)
tokenized_datasets = dataset.map(preprocess_function, batched=True)

import numpy as np
import evaluate
metric = evaluate.load("sacrebleu")

def postprocess_text(preds, labels):
    preds = [pred.strip() for pred in preds]
    labels = [[label.strip()] for label in labels]
    return preds, labels

def compute_metrics(eval_preds):
    preds, labels = eval_preds
    if isinstance(preds, tuple):
        preds = preds[0]
    decoded_preds = tokenizer.batch_decode(preds, skip_special_tokens=True)

    labels = np.where(labels != -100, labels, tokenizer.pad_token_id)
    decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)

    decoded_preds, decoded_labels = postprocess_text(decoded_preds, decoded_labels)

    result = metric.compute(predictions=decoded_preds, references=decoded_labels)
    result = {"bleu": result["score"]}

    prediction_lens = [np.count_nonzero(pred != tokenizer.pad_token_id) for pred in preds]
    result["gen_len"] = np.mean(prediction_lens)
    result = {k: round(v, 4) for k, v in result.items()}
    return result

training_args = Seq2SeqTrainingArguments(
    output_dir="chkpt3",
    learning_rate=2e-5,
    logging_dir='logs1',
    logging_steps=1,
    weight_decay=0.01,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    num_train_epochs=5,
    save_steps=10000,
    save_total_limit=50,
    evaluation_strategy="epoch",
    predict_with_generate=True,
)

trainer = Seq2SeqTrainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_datasets["train"],
    eval_dataset=tokenized_datasets["test"],
    tokenizer=tokenizer,
    data_collator=data_collator,
    compute_metrics=compute_metrics,
)

trainer.train()
trainer.save_model("result3")

wandb.finish()

이렇게 사용하고 있습니다.

**Comment by juhwanlee-diquest** [comment]
발생하셨던 문제점이 어떤 것이였을까요?

**Comment by kht6163** [comment]
3090 2개로 사용하고 있는데 gram를 굉장히 많이 먹더라구요 배치사이즈를 8이상으로 설정하면 메모리 부족으로 학습하다가 꺼져 버리더라구요

그리고 하이퍼 파라미터 값도 저렇게하면 적당한지 궁금합니다.

**Comment by juhwanlee-diquest** [comment]
저희가 학습시 사용한 GPU와 하이퍼파라미터 값입니다.

GPU: A100-80G

batch_size = 32
learning_rate=2e-5
fp16=True
weight_decay=0.01,
epoch=3

**Comment by kht6163** [comment]
아 그리고 혹시 추가로 궁금한게 한-영 영-한 모델을 각각 따로 학습 시킨 이유가 있으신가요?

**Comment by juhwanlee-diquest** [comment]
저희는 데이터 셋에 초점을 맞추었기에  분류했었습니다.

영한 데이터셋과 한영데이터 셋은 다르다고 판단하고 진행했기 때문입니다.

데이터 셋 목적이 영어를 한국어로 바꾼것과, 한국어를 영어로 바꾼 것은 차이가 있을 것이다라는 생각을 가정하고 진행했습니다.
